pkg/cache/cache_test.go

```diff
@@ -0,0 +1,871 @@
+	}, map[string]cache.Interface[[]byte]{
+		"source1": source1,
+		"source2": source2,
+	})
+	data, etag, err := merger.Data().Split()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if want := "merged source1 and source2"; string(data) != want {
+		t.Fatalf("expected data = %v, got %v", want, string(data))
+	}
+	if want := "merged source1 and source2"; etag != want {
+		t.Fatalf("expected etag = %v, got %v", want, etag)
+	}
+	_, _, err = merger.Data().Split()
+	if err == nil {
+		t.Fatalf("expected error, none found")
+	}
+	data, etag, err = merger.Data().Split()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if want := "merged source1 and source2"; string(data) != want {
+		t.Fatalf("expected data = %v, got %v", want, string(data))
+	}
+	if want := "merged source1 and source2"; etag != want {
+		t.Fatalf("expected etag = %v, got %v", want, etag)
+	}
+	_, _, err = merger.Data().Split()
+	if err == nil {
+		t.Fatalf("expected error, none found")
+	}
+	if source1Count != 4 {
+		t.Fatalf("Expected source function called 4x, called: %v", source1Count)
+	}
+	if source2Count != 4 {
+		t.Fatalf("Expected source function called 4x, called: %v", source2Count)
+	}
+	if mergerCount != 4 {
+		t.Fatalf("Expected merger function called 4x, called: %v", mergerCount)
+	}
+}
+
+func TestReplaceable(t *testing.T) {
+	sourceDataCount := 0
+	sourceData := cache.NewSource(func() cache.Result[[]byte] {
+		sourceDataCount += 1
+		return cache.NewResultData([]byte("source"), "source")
+	})
+	sourceData2Count := 0
+	sourceData2 := cache.NewSource(func() cache.Result[[]byte] {
+		sourceData2Count += 1
+		return cache.NewResultData([]byte("source2"), "source2")
+	})
+
+	sourceErrCount := 0
+	sourceErr := cache.NewSource(func() cache.Result[[]byte] {
+		sourceErrCount += 1
+		return cache.NewResultError[[]byte](errors.New("source error"))
+	})
+	replaceable := cache.Replaceable[[]byte]{}
+	replaceable.Replace(sourceErr)
+	_, _, err := replaceable.Data().Split()
+	if err == nil {
+		t.Fatalf("expected error, found none")
+	}
+	replaceable.Replace(sourceData)
+	data, etag, err := replaceable.Data().Split()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if want := "source"; string(data) != want {
+		t.Fatalf("expected data = %v, got %v", want, string(data))
+	}
+	if want := "source"; etag != want {
+		t.Fatalf("expected etag = %v, got %v", want, etag)
+	}
+	// replace with the same thing, shouldn't change anything
+	replaceable.Replace(sourceData)
+	data, etag, err = replaceable.Data().Split()
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if want := "source"; string(data) != want {
+		t.Fatalf("expected data = %v, got %v", want, string(data))
+	}
+	if want := "source"; etag != want {
+		t.Fatalf("expected etag = %v, got %v", want, etag)
+	}
+	// Even if we replace with something that fails, we continue to return the success.
```

Comment:
How do we surface the error in this case?

Reply:
We don't, I could have a different type that doesn't do that, but it's pretty common, at least in the openapi business, that we continue to return stale data rather than new errors.